[Parse]: Separate do and while blocks diagnostics

kitaisreal opened the pull request:
SR-11148: Separate do and while blocks generate error from legacy diagnostic

1. Taking new line into account if there is a `while` token after stmt.
2. Updated diagnostic to provide new line separator.

More information in ticket https://bugs.swift.org/browse/SR-11148.

slavapestov committed:
Sema: Fix a couple of problems with capture analysis and TopLevelCodeDecls

A TopLevelCodeDecl is a local context and any declarations inside
of one must be treated as captures. Furthermore, all the
TopLevelCodeDecl children of a source file are peers, and can
see each other's bindings, so don't perform an exact match on the
DeclContext.

Part of <rdar://problem/23051362> / <https://bugs.swift.org/browse/SR-3528>.

slavapestov committed:
SILGen: Diagnose invalid captures of top-level bindings

A method of a nominal type is not allowed to capture anything.
If we did compute a capture list, clear it out to preserve
downstream invariants.

This can happen because not all invalid captures involving
nominal types in local context can be diagnosed in Sema.

In particular, we allow a method of a local type to
reference a local function from an outer scope, as long as
that local function does not have any (transitive) captures.
Since Sema cannot validate the latter condition, it punts
and allows any reference to a FuncDecl in local context.

When emitting the captures in this situation, SILGen will
diagnose the problem using the existing 'forward reference'
logic.

The diagnostic is sub-optimal -- it should talk about a
reference from inside a local type, and not a 'forward
reference'. But still better than crashing.

slavapestov committed:
SILGen: Clean up invariants around capture lists

kitaisreal committed:
SR-11148: Separate do and while blocks generate error from legacy diagnostic
1. Taking new line into account if there is a `while` token after stmt.
2. Updated diagnostic to provide new line separator.

CodaFi commented:
Could you please add tests?

rintaro commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -1917,18 +1917,23 @@ ParserResult<Stmt> Parser::parseStmtDo(LabeledStmtInfo labelInfo) {
   diagnose(doLoc, diag::do_while_now_repeat_while)
-    .fixItReplace(doLoc, "repeat");
+    .fixItReplace(doLoc, "repeat")
+    .fixItInsert(whileLoc, "\n");
```
I don't think this is right.
If the user applies these fix-its, the code will be something like:
```swift
repeat {
}
while condition
```
The newline is not needed at all.

If we want users to choose between `repeat` and bare `do`, you need to emit 3 diagnostics:
```
error: 'do-while' statement is not allowed.
note: did you mean 'repeat-file' statement?  (with fix-it to 'repeat')
note: did you mean separate 'do' and 'while' statements? (with fix-it inserting '\n')
```

rintaro commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -1917,18 +1917,23 @@ ParserResult<Stmt> Parser::parseStmtDo(LabeledStmtInfo labelInfo) {
-
-  // But if we do, advise the programmer that it's 'repeat' now.
+  SourceLoc whileLoc = Tok.getLoc();
+  if (Tok.isAtStartOfLine()) {
```
Could you cosolidate these two `if` into one?
```cpp
if (Tok.getKind() != tok::kw_while || Tok.isStartOfLine()) {
    return makeParserResult(...);
}

rintaro reviewed (changes requested):
Thank you for working on this! Just a couple of requests:

slavapestov committed:
Sema: Tweak SanitizeExpr handling of AutoClosureExprs

Curry thunks create AutoClosureExprs with parameters, and we can't just
fold those away.

Also, remove an older AutoClosureExpr cleanup that seems to be
redundant.

slavapestov committed:
Sema: Remove TypeCheckExprFlags::SkipMultiStmtClosures

Fold this into TypeCheckExprFlags::SubExpressionDiagnostics, which
is itself going away soon, hopefully!

slavapestov committed:
Revert "SILGen: Simplify SILGenFunction::emitClosure()"

This change has a subtle impact on debug info emission and causes
a regression with a subsequent patch I'm about to commit.

This reverts commit 62d1adb409918c638b53187cfb6ae3f4380afbd6.

slavapestov committed:
SILOptimizer: AllocBoxToStack preserves [transparent] bit in cloned function

We need this to uphold the invariant that in the performance pipeline
before SIL serialization occurs, a function has ownership iff it is
transparent.

slavapestov committed:
AST: Simplify Expr::getReferencedDecl()

slavapestov committed:
IDE: Preparations for Sema building curry thunks

slavapestov committed:
Parse: Remove some dead code

slavapestov committed:
Parse: Parse entire top level in one shot

slavapestov committed:
Parse: Keep track of top-level declarations that follow a 'guard' statement

compnerd committed:
Driver: mark test as XFAIL on windows

slavapestov committed:
AST: More consistent definition of a 'local capture'

This commit adds a new ValueDecl::isLocalCapture() predicate and
uses it in the right places. The predicate is true if the
declaration is in local context, *or* if its at the top level of
the main source file and follows a 'guard' statement.

Fixes <rdar://problem/23051362> / <https://bugs.swift.org/browse/SR-3528>.

rintaro committed:
[CodeCompletion] Eliminate CodeCompletionCallbacks::completeExpr()

Well, this does nothing. It just set the parser position twice and calls
the consumer's 'handleResults()' without any reason.
It seems it was intended to be a "fallback" completion, but it was never
implemented properly. So just remove it for now.

rdar://problem/58102910

Commit:
Make path suitable for Windows

theblixguy committed:
[Typechecker] Fix duplicate diagnostics when using attributes (#28888)

* [Typechecker] resolveAttributedType() should invalidate the type repr after emitting a diagnostic, to prevent duplicate diagnostics later

* [Test] Update '@noescape' attribute diagnostics

* [Typechecker] Add a special diagnose method to TypeResolver that accepts a TypeRepr and invalidates it

* [Typechecker] Rename the special 'diagnose' method to 'diagnoseInvalid' for clarity

drodriguez committed:
[MSVC] Evaluate parameters in the right order.

Parameter evaluation order is unspecified, and MSVC happens to do it in
the opposite order than Clang.

Move the evaluation of the parameters outside the invocation, so the
order is clear, and the token consuming happens in the right order.

CodaFi committed:
Remove re-entrancy assertion

This was just to confirm that the re-entrancy problem still exists on
the bots.  There's something OS-dependent (probably the SDKs) that
causes this to only reproduce on 10.14.5 and not 10.15.

rdar://58116531

compnerd committed:
Revert "Driver: mark test as XFAIL on windows"

kitaisreal committed:
Merge branch 'master' of https://github.com/apple/swift
